plugins/utils/todo-comments.nix

```diff
@@ -16,275 +24,400 @@ let
-                example = "~/projects/foobar";
-              };
-
-              keywords = mkOption {
-                type = types.nullOr types.str;
-                description = ''
-                  Comma separated list of keywords to filter results by.
-                  Keywords are case-sensitive.
-                '';
-                default = null;
-                example = "TODO,FIX";
-              };
-
-              options = helpers.keymaps.mapConfigOptions;
-            };
-        in
-        mapAttrs mkKeymapOption commands;
     };
   };
 
-  config =
-    let
-      setupOptions = {
-        inherit (cfg) signs;
-        sign_priority = cfg.signPriority;
-        inherit (cfg) keywords;
-        gui_style = cfg.guiStyle;
-        merge_keywords = cfg.mergeKeywords;
-        highlight = {
-          inherit (cfg.highlight)
-            multiline
-            before
-            keyword
-            after
-            exclude
-            ;
-          pattern = helpers.ifNonNull' cfg.highlight.pattern (helpers.mkRaw "[[${cfg.highlight.pattern}]]");
-          multiline_pattern = cfg.highlight.multilinePattern;
-          multiline_context = cfg.highlight.multilineContext;
-          comments_only = cfg.highlight.commentsOnly;
-          max_line_len = cfg.highlight.maxLineLen;
-        };
-        inherit (cfg) colors;
-        search = {
-          inherit (cfg.search) command args;
-          pattern = helpers.ifNonNull' cfg.search.pattern (
-            if isList cfg.search.pattern then
-              (map helpers.mkRaw cfg.search.pattern)
-            else
-              helpers.mkRaw "[[${cfg.search.pattern}]]"
-          );
-        };
-      } // cfg.extraOptions;
-    in
-    mkIf cfg.enable {
-      extraPlugins = [ cfg.package ];
-      extraPackages = [ cfg.ripgrepPackage ];
-      extraConfigLua = ''
-        require("todo-comments").setup${helpers.toLuaObject setupOptions}
-      '';
+  # TODO: any good examples?
+  settingsExample = {
 
-      keymaps = flatten (
-        mapAttrsToList (
+  };
+
+  extraOptions = {
+    keymaps =
+      let
+        mkKeymapOption =
           optionName: funcName:
-          let
-            keymap = cfg.keymaps.${optionName};
-
-            cwd = optionalString (keymap.cwd != null) " cwd=${keymap.cwd}";
-            keywords = optionalString (keymap.keywords != null) " keywords=${keymap.keywords}";
-          in
-          optional (keymap != null) {
-            mode = "n";
-            inherit (keymap) key options;
-            action = ":${funcName}${cwd}${keywords}<CR>";
-          }
-        ) commands
-      );
-
-      # Automatically enable plugins if keymaps have been set
-      plugins = mkMerge [
-        (mkIf (cfg.keymaps.todoTrouble != null) { trouble.enable = true; })
-        (mkIf (cfg.keymaps.todoTelescope != null) { telescope.enable = true; })
-      ];
+          mkCompositeOption "Keymap settings for the `:${funcName}` function." {
```

Comment:
Is it possible to do this is a way that complies with https://github.com/nix-community/nixvim/issues/603?

Reply:
If you're unfamiliar, the idea is that plugin keymaps should make use of `mkMapOptionSubmodule` instead of implementing their own keymap submodule types.

Reply:
If `mkMapOptionSubmodule` is too limiting, because it doesn't support additional options (`cwd`, `keywords`) I'm open to making it more powerful by adding an `extraOptions` argument or similar.

Reply:
As discussed in https://github.com/nix-community/nixvim/pull/2030#discussion_r1720865792 this can done in be another PR